design/core_collection_operations.md

```diff
@@ -0,0 +1,175 @@
+  * The version information allows creating a new version of the collection document but only
+    updating some of the associated data products, while leaving others at their current
+    version
+  * Similar to the collections version, it also allows for easy rollbacks, and any version
+    semantics are up to the user.
+  * In the future, we could add a schema version if the schema needs to change. For now, no
+    schema version implies a schema version of 1.
+    * An altenative would just be to add a version to the product name, like `taxa_freq2`.
+
+```
+[
+    {
+        "product": "taxa_freq",
+        "version": "r207.kbase.3"
+    },
+    {
+        "product": "genome_stats",
+        "version': "r207.kbase.14"
+    }
+]
+```
+
+## Implementation
+
+### DB schema
+
+* The collections document will be essentially as represented above, with the keys / values
+  as described and a list of associated data products.
+* There will be two separate collections - `coll_all_versions` (or something) and
+  `coll_active_versions`.
+  * The former holds all versions of all collections, and the latter holds only the active
+    version of each collection.
+* Activating a version means:
+  * Pulling the collection version document from `coll_all_versions`
+  * Setting the activation date to the current date
+  * Overwriting the collection document in `coll_active_versions`.
+    * Atomic!
+* Alternatives:
+  * Tag the active version in `coll_active_version`.
+    * There's a fundamental problem with this approach in that it's non-atomic - at some point
+      either no versions will be tagged as active or > 1 will.
+    * This can be alleviated with transactions but in Arango [transactions are only reliable in
+      non-sharded environments](https://github.com/arangodb/arangodb/issues/11424).
+    * Furthermore, filtering data becomes much slower since all the inactive versions need to
+      be filtered out. Index size goes up because you need (at least) the activity state in
+      the equality part of the index. If you want to filter inactive versions, you need a
+      separate index without the activity state in the equality part of the index.
+    * Also generally more complex than the proposed schema.
+  * Have the service manage the collection version with a monotonically increasing version, like
+    `workspace_deluxe` or `sample_service`
+    * This approach is much more complex and is not necessary for the use case proposed here. It
+      also has the filtering drawback as mentioned above.
+
+### Endpoints
+
+* Only includes immediately relevant endpoints.
+
+#### List collections
+
+* For now returns all active collections based on the < 1000 collections assumption.
+
+```
+GET /collections
+```
+
+#### Save a collection
+
+* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.
+
+```
+POST /collections/{collection_id}/versions/{version}
```

Comment:
`PUT` _is_ actually more appropriate here, as `GET /collections/{collection_id}/versions/{version}` would be our likely way to fetch a specific collection version, and restful `POST`s usually generate IDs/resource-locations whereas `PUT` typically explicitly specify them

Reply:
This operation isn't idempotent, so a PUT isn't correct here

Reply:
This looks pretty idempotent to me, we should discuss

Reply:
see https://stackoverflow.com/questions/51953943/what-is-idempotency-in-http-put-can-i-disallow-overwriting-of-a-resource

unless I'm misunderstanding what the endpoint does

Reply:
Huh, TIL. https://developer.mozilla.org/en-US/docs/Glossary/Idempotent

That doesn't match with the standard CS definition of idempotent as I understand it, which is confusing if so.

Fair enough then.

Reply:
Yeah... AFAIK web folks tend to use idempotent to mean `idempotent with respect to the recipient` and ignore the existence of the sender

Reply:
resolving per @dauglyon